hibernate-core/src/main/java/org/hibernate/query/sqm/tree/predicate/SqmMemberOfPredicate.java

```diff
@@ -34,9 +39,11 @@ public SqmMemberOfPredicate(
 		this.pluralPath = pluralPath;
 		this.leftHandExpression = leftHandExpression;
 
-		leftHandExpression.applyInferableType(
-				( (SqmPluralValuedSimplePath<?>) pluralPath ).getReferencedPathSource().getElementType()
-		);
+		final SimpleDomainType simpleDomainType = ( (SqmPluralValuedSimplePath<?>) pluralPath ).getReferencedPathSource().getElementType();
```

Comment:
IMO you should change the parameter type of `pluralPath` to `SqmPluralValuedSimplePath` to avoid the cast at this point. All callers already do the type check, so we might as well pass the correct type. Also change the field type of `pluralPath`.